Mathlib/Analysis/Complex/UpperHalfPlane/Manifold.lean

```diff
@@ -31,4 +32,31 @@ theorem smooth_coe : Smooth 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ → ℂ) := fun _ =
 theorem mdifferentiable_coe : MDifferentiable 𝓘(ℂ) 𝓘(ℂ) ((↑) : ℍ → ℂ) :=
   smooth_coe.mdifferentiable
 
+lemma smoothAt_ofComplex {z : ℂ} (hz : 0 < z.im) :
+    SmoothAt 𝓘(ℂ) 𝓘(ℂ) ofComplex z := by
+  rw [SmoothAt, contMDiffAt_iff]
+  constructor
+  · -- continuity at z
+    rw [ContinuousAt, nhds_induced, tendsto_comap_iff]
+    refine Tendsto.congr' (eventuallyEq_coe_comp_ofComplex hz).symm ?_
+    simpa only [ofComplex_apply_of_im_pos hz, Subtype.coe_mk] using tendsto_id
+  · -- smoothness in local chart
+    simp only [extChartAt, PartialHomeomorph.extend, modelWithCornersSelf_partialEquiv,
+      PartialEquiv.trans_refl, PartialHomeomorph.toFun_eq_coe, PartialHomeomorph.refl_partialEquiv,
+      PartialEquiv.refl_source, PartialHomeomorph.singletonChartedSpace_chartAt_eq,
+      PartialEquiv.refl_symm, PartialEquiv.refl_coe, CompTriple.comp_eq, modelWithCornersSelf_coe,
+      Set.range_id, id_eq, contDiffWithinAt_univ]
+    exact contDiffAt_id.congr_of_eventuallyEq (eventuallyEq_coe_comp_ofComplex hz)
+
+lemma mdifferentiableAt_ofComplex {z : ℂ} (hz : 0 < z.im) :
+    MDifferentiableAt 𝓘(ℂ) 𝓘(ℂ) ofComplex z :=
+  (smoothAt_ofComplex hz).mdifferentiableAt
+
+lemma mdifferentiableAt_iff {f : ℍ → ℂ} {τ : ℍ} :
```

Comment:
while we are here, do we want something like 

```
lemma mdifferentiable_iff {f : ℍ → ℂ} {τ : ℍ} :
    MDifferentiable 𝓘(ℂ) 𝓘(ℂ) f  ↔ DifferentiableOn ℂ (f ∘ ofComplex) {z : ℂ | 0 < z.im} := by ..
```

Since MDifferentiable is what shows up in the modular form definition, so this would be a nice way to get the condition directly?

Reply:
I'm not sure this would be so helpful, because the `DifferentiableOn` doesn't mean "`f` is differentiable at every point of `s`" but "`f` is differentiable within `s` at every point of `s`". If `s` is open these conditions are equivalent but one needs to invoke a lemma to show this, which kind of cancels out the usability gain. 

It's a pity we don't have a direct way of asserting "f is differentiable at every point of `s`", analogous to `AnalyticOnNhds`.

Reply:
On second thoughts, it can't do any harm to have this lemma as well. I've added it.